Parse/Internal/File/Controller/PFFileStagingController.m

```diff
@@ -0,0 +1,105 @@
+    PFAsyncTaskQueue *_taskQueue;
+}
+
+///--------------------------------------
+#pragma mark - Init
+///--------------------------------------
+
+- (instancetype)init {
+    PFNotDesignatedInitializer();
+}
+
+- (instancetype)initWithDataSource:(id<PFFileManagerProvider>)dataSource {
+    self = [super init];
+    if (!self) return nil;
+
+    _dataSource = dataSource;
+    _taskQueue = [PFAsyncTaskQueue taskQueue];
+
+    [self _clearStagedFiles];
+
+    return self;
+}
+
++ (instancetype)stagingControllerWithDataSource:(id<PFFileManagerProvider>)dataSource {
+    return [[self alloc] initWithDataSource:dataSource];
+}
+
+///--------------------------------------
+#pragma mark - Properties
+///--------------------------------------
+
+- (NSString *)stagedFilesDirectoryPath {
+    NSString *folderPath = [self.dataSource.fileManager parseLocalSandboxDataDirectoryPath];
+    return [folderPath stringByAppendingPathComponent:PFFileStagingControllerDirectoryName_];
+}
+
+///--------------------------------------
+#pragma mark - Staging
+///--------------------------------------
+
+- (BFTask *)stageFileAtPath:(NSString *)filePath name:(NSString *)name uniqueId:(uint64_t)uniqueId {
+    return [_taskQueue enqueue:^id(BFTask *task) {
+        NSString *destinationPath = [self stagedDirectoryPathForFileWithName:name uniqueId:uniqueId];
+        NSFileManager *fileManager = [NSFileManager defaultManager];
+
+        NSError *error = nil;
+        if (![fileManager copyItemAtPath:filePath toPath:destinationPath error:&error]) {
+            return [BFTask taskWithError:error];
+        }
+
+        return [BFTask taskWithResult:destinationPath];
+    }];
+}
+
+- (BFTask *)stageFileWithData:(NSData *)fileData name:(NSString *)name uniqueId:(uint64_t)uniqueId {
+    return [_taskQueue enqueue:^id(BFTask *task) {
+        NSString *destinationPath = [self stagedDirectoryPathForFileWithName:name uniqueId:uniqueId];
+        NSError *error = nil;
+        if (![fileData writeToFile:destinationPath options:NSDataWritingAtomic error:&error]) {
+            return [BFTask taskWithError:error];
+        }
+
+        return [BFTask taskWithResult:destinationPath];
+    }];
+}
+
+- (NSString *)stagedDirectoryPathForFileWithName:(NSString *)name uniqueId:(uint64_t)uniqueId {
+    NSString *fileName = [NSString stringWithFormat:@"%llX_%@", uniqueId, name];
+    return [[self stagedFilesDirectoryPath] stringByAppendingPathComponent:fileName];
+}
+
+///--------------------------------------
+#pragma mark - Clearing
+///--------------------------------------
+
+- (BFTask *)_clearStagedFiles {
+    return [_taskQueue enqueue:^id(BFTask *task) {
+        NSString *stagedFilesDirectoryPath = [self stagedFilesDirectoryPath];
+        return [[PFFileManager removeItemAtPathAsync:stagedFilesDirectoryPath] continueWithBlock:^id(BFTask *task) {
+            return [PFFileManager createDirectoryIfNeededAsyncAtPath:stagedFilesDirectoryPath];
```

Comment:
What do you think if we would ditch this directory creation here and rather check if directory exists whenever staging new files? If clear didn't happen (say we remove this functionality later or the folder is simply gone) - we will fail with an error, while it should actually recreate that folder.